File: app/app/src/main/java/me/williamhester/reddit/inject/ActivityModule.java
package me.williamhester.reddit.inject;

import com.google.gson.JsonParser;

import org.greenrobot.eventbus.EventBus;

import java.io.IOException;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import me.williamhester.reddit.apis.RedditClient;
import me.williamhester.reddit.apis.RedditHttpRequestFactory;
import me.williamhester.reddit.convert.RedditGsonConverter;
import me.williamhester.reddit.models.managers.AccountManager;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/** A module for creating Activity-wide instances of classes */
@Module
public class ActivityModule {
  @Provides
  @Singleton
  EventBus provideBus() {
    return new EventBus();
  }

  @Provides
  @Singleton
  OkHttpClient provideOkClient() {
    Interceptor headerInterceptor = new Interceptor() {
      @Override
      public Response intercept(Chain chain) throws IOException {
        Request original = chain.request();

        // Customize the request
        Request.Builder builder = original.newBuilder()
            .addHeader("User-Agent", "Breadit-2")
            .addHeader("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8")
            .method(original.method(), original.body());

        return chain.proceed(builder.build());
      }
    };

    return new OkHttpClient.Builder()
        .addInterceptor(headerInterceptor)
        .build();
  }

  @Provides
  @Singleton
  RedditHttpRequestFactory provideHttpRedditRequestFactory(
      OkHttpClient client,
      EventBus bus,
      AccountManager accountManager,
      JsonParser parser) {
    return new RedditHttpRequestFactory(client, accountManager, bus, parser);
  }

  @Provides
  @Singleton
  RedditClient provideRedditApi(RedditGsonConverter converter, RedditHttpRequestFactory factory) {
    return new RedditClient(converter, factory);
  }
}

File: app/app/src/main/java/me/williamhester/reddit/inject/ActivityModule.kt
package me.williamhester.reddit.inject

import com.google.gson.JsonParser
import dagger.Module
import dagger.Provides
import me.williamhester.reddit.apis.RedditClient
import me.williamhester.reddit.apis.RedditHttpRequest
import me.williamhester.reddit.apis.RedditHttpRequestExecutor
import me.williamhester.reddit.convert.RedditGsonConverter
import me.williamhester.reddit.html.HtmlParser
import me.williamhester.reddit.models.managers.AccountManager
import me.williamhester.reddit.ui.ContentClickCallbacks
import me.williamhester.reddit.ui.ContentClickCallbacksImpl
import okhttp3.OkHttpClient
import org.greenrobot.eventbus.EventBus

/**
 * A module that defines all of the dependencies that should live as long as an Activity is alive.
 */
@Module
class ActivityModule {
  @Provides
  @ActivityScope
  internal fun provideBus(): EventBus {
    return EventBus()
  }

  @Provides
  @ActivityScope
  internal fun provideHttpRedditRequestFactory(
      client: OkHttpClient,
      bus: EventBus,
      accountManager: AccountManager,
      parser: JsonParser,
      redditGsonConverter: RedditGsonConverter): RedditHttpRequestExecutor {
    return RedditHttpRequestExecutor(client, accountManager, parser, redditGsonConverter)
  }

  @Provides
  @ActivityScope
  internal fun provideRedditApi(
      converter: RedditGsonConverter,
      executor: RedditHttpRequestExecutor,
      requestFactory: RedditHttpRequest.Builder.Factory,
      bus: EventBus): RedditClient {
    return RedditClient(converter, requestFactory, bus)
  }

  @Provides
  @ActivityScope
  internal fun provideVotableCallbacks(bus: EventBus): ContentClickCallbacks {
    return ContentClickCallbacksImpl(bus)
  }

  @Provides
  @ActivityScope
  internal fun provideHtmlParser(callbacks: ContentClickCallbacks): HtmlParser {
    return HtmlParser(callbacks)
  }

  @Provides
  @ActivityScope
  internal fun provideRedditHttpRequestBuilderFactory(
      accountManager: AccountManager,
      requestExecutor: RedditHttpRequestExecutor): RedditHttpRequest.Builder.Factory {
    return RedditHttpRequest.Builder.Factory(accountManager, requestExecutor)
  }
}